### PR TITLE
[MENFORCER-467] banDynamicVersions excludedScopes on project level

### PR DESCRIPTION
**Pull request by slawekjaranowski**
We should not change default Maven scope filtering for transitive dependencies

---

Put message of violation into exception to be consistent with other rules

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MENFORCER) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MENFORCER-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MENFORCER-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

File: enforcer-rules/src/main/java/org/apache/maven/enforcer/rules/dependency/BanDynamicVersions.java [hunk 1]
```diff
@@ -138,19 +124,19 @@ private final class BannedDynamicVersionCollector implements DependencyVisitor {
...
         private boolean isRoot = true;
 
-        private int numViolations;
+        private List<String> violations;
 
         private final Predicate<DependencyNode> predicate;
 
-        public int getNumViolations() {
-            return numViolations;
+        public List<String> getNumViolations() {
```
Comment (kwin) [inline enforcer-rules/src/main/java/org/apache/maven/enforcer/rules/dependency/BanDynamicVersions.java:131]:
```suggestion
        public List<String> getViolations() {
```

File: enforcer-rules/src/main/java/org/apache/maven/enforcer/rules/dependency/BanDynamicVersions.java [hunk 2]
```diff
@@ -256,10 +231,8 @@ public boolean test(DependencyNode depNode) {
         }
     }
 
-    private int emitDependenciesWithBannedDynamicVersions(
-            Dependency rootDependency, RepositorySystemSession repoSession) throws DependencyCollectionException {
-        CollectRequest collectRequest = new CollectRequest(rootDependency, project.getRemoteProjectRepositories());
-        CollectResult collectResult = repoSystem.collectDependencies(repoSession, collectRequest);
+    private List<String> emitDependenciesWithBannedDynamicVersions(DependencyNode rootDependency)
```
Comment (kwin) [inline enforcer-rules/src/main/java/org/apache/maven/enforcer/rules/dependency/BanDynamicVersions.java:234]:
```suggestion
    private List<String> collectDependenciesWithBannedDynamicVersions(DependencyNode rootDependency)
```

File: enforcer-rules/src/main/java/org/apache/maven/enforcer/rules/dependency/ResolverUtil.java [hunk 3]
```diff
@@ -96,6 +96,19 @@ DependencyNode resolveTransitiveDependencies() throws EnforcerRuleException {
         return resolveTransitiveDependencies(false, true, Arrays.asList(SCOPE_TEST, SCOPE_PROVIDED));
     }
 
+    /**
+     * Retrieves the {@link DependencyNode} instance containing the result of the transitive dependency
+     * for the current {@link MavenProject}.
+     *
+     * @param excludedScopes a project dependency scope to excluded
```
Comment (kwin) [inline enforcer-rules/src/main/java/org/apache/maven/enforcer/rules/dependency/ResolverUtil.java:103]:
```suggestion
     * @param excludedScopes the scopes of direct dependencies to ignore
```

File: enforcer-rules/src/main/java/org/apache/maven/enforcer/rules/dependency/ResolverUtil.java [hunk 4]
```diff
@@ -96,6 +96,19 @@ DependencyNode resolveTransitiveDependencies() throws EnforcerRuleException {
...
+     * @return a Dependency Node which is the root of the project's dependency tree
+     * @throws EnforcerRuleException thrown if the lookup fails
+     */
+    DependencyNode resolveTransitiveDependencies(boolean excludeOptional, List<String> excludedScopes)
```
Comment (kwin) [inline enforcer-rules/src/main/java/org/apache/maven/enforcer/rules/dependency/ResolverUtil.java:107]: the javadoc for `excludeOptional` is missing

**Review by kwin — CHANGES_REQUESTED** [review]

**Review by kwin — APPROVED** [review]

**Comment by jira-importer** [comment]
Resolve #594

**Pull request merged**